### PR TITLE
Import Employers and Clients from CSV

### LINKED ISSUE #140: Import Clients from CSV
# Design

I dropped the ball, didn't mention this early enough to designers. If you get assigned this, msg me + designers about if the design for the import is there yet

# Implementation
- [ ] Create endpoint, which given CSV file (msg me for access to the file with their data) and creates a Client for each of them
- [ ] implement whatever frontend there is

# Notes:
- This may not adhere to the schema, in terms of missing data. I'm fine with loosening our requirements, and just encoding them in the form checking + backend body checking, that way we can import them into our system here.

### LINKED ISSUE #141: Import Employers from a CSV
# Design

I dropped the ball, didn't mention this early enough to designers. If you get assigned this, msg me + designers about if the design for the import is there yet

# Implementation
- [ ] Create endpoint, which given CSV file (msg me for access to the file with their data) and creates an Employer for each of them
- [ ] implement whatever frontend there is

# Notes:
- This may not adhere to the schema, in terms of missing data. I'm fine with loosening our requirements, and just encoding them in the form checking + backend body checking, that way we can import them into our system here.

### PR DESCRIPTION
**Pull request by helen-li**
# Description

* Client UI for simple file uploading (no official design, make it look nicer if time)
* Server endpoints for importing data from csv files into the client and employer collections
* 2 new dependencies added: fast-csv (converting from csv to json), multer (file uploading)

Screen Shot 2023-04-27 at 2 54 26 PM

Fixes #140, Fixes #141 

## Sample curl
curl --location --request POST 'http://localhost:3001/api/employer/import/csv' (you also need to upload a file in request)
## Response
A collection of object IDs of all the newly-imported items

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have run `npm run lint:fix`
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if required)
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works (Delete if frontend)

### CODE REVIEW

File: client/src/components/UploadFile/UploadFile.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      const formData = new FormData();
+      formData.append("file", file);
+
+      fetch(`/api/${selectedOption}/import/csv`, {
+        method: "POST",
+        body: formData,
+      })
+        .then((response) => response.json())
+        .then((data) => {
+          console.log(`Imported ${data.insertedCount} to client collection: ${data.insertedIds}.`);
```
Comment (kenneth-miura) [inline client/src/components/UploadFile/UploadFile.tsx:42]: can u make this an `alert` so it pops up for user?

Also check the response code, and throw an error if it's not a success code.

Then put a .catch that uses `alert` to say the operation failed

File: server/src/models/client.model.ts [hunk 2]
```diff
@@ -444,6 +446,49 @@ export default class Client {
...
+            .on('end', () => {
+                result.shift();   // remove the first line: header
+                resolve(result);
+            })
+            .on('error', (error: Error) => {
+                reject(error);
+            });
+            stream.pipe(csvStream);
+        });
+        // debug(jsonArray);
```
Comment (kenneth-miura) [inline server/src/models/client.model.ts:481]: can u delete this comment

Comment (kenneth-miura) [reply server/src/models/client.model.ts:481]: this is like commenting out code right? not docs?

File: server/src/models/employer.model.ts [hunk 3]
```diff
@@ -300,6 +302,68 @@ export default class Employer {
...
+            stream.pipe(csvStream);
+        });
+        const addressResult = await collections.addresses?.insertMany(jsonArray.map((doc) => doc.address));
+        if (!addressResult) {
+            throw new Error('Failed to import to employer collection during address insertions.')
+        }
+        for (let i = 0; i < jsonArray.length; i++) {
+            jsonArray[i].employer.primary_address_id = addressResult.insertedIds[i];
+        }
+        // debug(jsonArray);
```
Comment (kenneth-miura) [inline server/src/models/employer.model.ts:356]: same as before

**Review by kenneth-miura — CHANGES_REQUESTED** [review]
Looks great to me! 2 or 3 small changes, but besides that looks great! Thanks for the quick turn around

**Review by jordanjanakievski — APPROVED** [review]
Apart from Kenneth's comments, everything looks good. I definitely second the alert so the user knows whether the upload was successful or unsuccessful. It was not happy when I tried sending it a pdf.

File: client/src/components/UploadFile/UploadFile.tsx (hunk 1 above)
Comment (helen-li) [reply client/src/components/UploadFile/UploadFile.tsx:42]: good call will do

**Review by kenneth-miura — APPROVED** [review]
GODLIKE PR

**Pull request merged**